src/entry.rs

```diff
@@ -152,43 +171,86 @@ pub(crate) fn gen_entry_point(input: &str) -> (String, String) {
 return Err(FutharkError::new(ctx).into());}}"
     );
 
+    let mut opaque_types = Vec::new();
     // OUTPUT
     let mut result_counter = 0;
     write!(&mut buffer2, "Ok(");
-    for (i, e) in entry_points.iter().enumerate() {
-        if e.1.starts_with("out") {
+    for (i, (argtype, argname)) in arg_pairs.iter().enumerate() {
+        if argname.starts_with("out") {
+            if !parse_array_type(argtype).is_some() {
+                opaque_types.push(argtype.clone());
+            }
             if result_counter > 0 {
                 write!(&mut buffer2, ", ");
             }
             result_counter += 1;
-            if e.0.starts_with("futhark") {
+            if argtype.starts_with("futhark") {
                 writeln!(
                     &mut buffer2,
                     "{}::from_ptr(ctx, raw_{})",
-                    auto_ctor(&e.0),
-                    e.1
+                    auto_ctor(&argtype),
+                    argname
                 );
             } else {
-                writeln!(&mut buffer2, "raw_{}", e.1);
+                writeln!(&mut buffer2, "raw_{}", argname);
             }
         }
     }
     write!(&mut buffer2, ")\n}}");
-    (buffer, buffer2)
+
+    (buffer, buffer2, opaque_types)
+}
+fn to_opaque_type_name(s: &str) -> String {
+    let mut rust_opaque_type = s.to_camel_case();
+
+    if let Some(r) = rust_opaque_type.get_mut(0..1) {
+        r.make_ascii_uppercase();
+    }
+    rust_opaque_type
+}
+
+fn gen_opaque_type(opaque_type: &str) -> String {
+    let rust_opaque_type = to_opaque_type_name(opaque_type);
+    let base_type = if opaque_type.starts_with("futhark_") {
+        &opaque_type[8..]
+    } else {
+        panic!("Apparent opaque type didn't start with futhark_.")
```

Comment:
Can you print the string that it is trying to make into an opaque type here as well, to make it easier to debug.

Reply:
This actually cannot happen now because `gen_opaque_type` will only be called on strings that start with `"futhark"`. This is an artifact of the discovery process. I'll remove it.

Reply:
Wait, sorry, that is a lie. There is still the possibility of a prefixed name that doesn't have an underscore. I'm trying to remember if there's any reason not to move the underscore into the guard or not. I think that will work.

Reply:
My last commit improves the check. I left an assertion, but that is not there to catch any specific known case. It is there to protect against an accidental refactor leading to the function being called on bad input.